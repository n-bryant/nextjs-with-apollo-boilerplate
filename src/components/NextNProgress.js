import React from "react";
import NProgress from "nprogress";
import Router from "next/router";
import PropTypes from "prop-types";

// implementation of nprogress that displays the progress of a loading state as a progress bar with a spinner
class NextNProgress extends React.Component {
  static defaultProps = {
    color: "red",
    startPosition: 0.3,
    stopDelayMs: 200,
    height: 3
  };

  timer = null;

  // set start position and then start progress
  routeChangeStart = () => {
    NProgress.set(this.props.startPosition);
    NProgress.start();
  };

  // stop the progress animation once the route change is complete
  routeChangeEnd = () => {
    clearTimeout(this.timer);
    this.timer = setTimeout(() => {
      NProgress.done(true);
    }, this.props.stopDelayMs);
  };

  render() {
    const { color, height } = this.props;

    return (
      <style jsx global>{`
        #nprogress {
          pointer-events: none;
        }
        #nprogress .bar {
          background: ${color};
          position: fixed;
          z-index: 1031;
          top: 0;
          left: 0;
          width: 100%;
          height: ${height}px;
        }
        #nprogress .peg {
          display: block;
          position: absolute;
          right: 0px;
          width: 100px;
          height: 100%;
          opacity: 1;
        }
        #nprogress .spinner {
          display: "block";
          position: fixed;
          z-index: 1031;
          top: 19px;
          left: 6px;
        }
        #nprogress .spinner-icon {
          width: 18px;
          height: 18px;
          box-sizing: border-box;
          border: solid 2px transparent;
          border-top-color: ${color};
          border-left-color: ${color};
          border-radius: 50%;
          -webkit-animation: nprogresss-spinner 400ms linear infinite;
          animation: nprogress-spinner 400ms linear infinite;
        }
        .nprogress-custom-parent {
          overflow: hidden;
          position: relative;
        }
        .nprogress-custom-parent #nprogress .spinner,
        .nprogress-custom-parent #nprogress .bar {
          position: absolute;
        }
        @-webkit-keyframes nprogress-spinner {
          0% {
            -webkit-transform: rotate(0deg);
          }
          100% {
            -webkit-transform: rotate(360deg);
          }
        }
        @keyframes nprogress-spinner {
          0% {
            transform: rotate(0deg);
          }
          100% {
            transform: rotate(360deg);
          }
        }
      `}</style>
    );
  }

  componentDidMount() {
    const { options } = this.props;

    // apply any additionally provided configurations
    if (options) {
      NProgress.configure(options);
    }

    // bind router events
    Router.events.on("routeChangeStart", this.routeChangeStart);
    Router.events.on("routeChangeComplete", this.routeChangeEnd);
    Router.events.on("routeChangeError", this.routeChangeEnd);
  }
}
NextNProgress.propTypes = {
  // the color to set the spinner/progress bar to
  color: PropTypes.string,
  // where the progress bar should start
  startPosition: PropTypes.number,
  // the delay before stopping the animation
  stopDelayMs: PropTypes.number,
  // additional nprogress options #[https://github.com/rstacruz/nprogress#configuration]
  options: PropTypes.object
};

export default NextNProgress;
