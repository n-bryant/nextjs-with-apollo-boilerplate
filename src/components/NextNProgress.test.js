import NextNProgress from "./NextNProgress";
import NProgress from "nprogress";

describe("NextNProgress", () => {
  let props = {
    color: "#fff",
    startPosition: 0.3,
    stopDelayMs: 200,
    height: 3
  };

  jest.spyOn(NProgress, "configure");
  jest.spyOn(NProgress, "set");
  jest.spyOn(NProgress, "start");
  jest.spyOn(NProgress, "done");

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<NextNProgress {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should not set additional configurations when the component mounts if options are not passed through props", () => {
    mount(<NextNProgress {...props} />);
    expect(NProgress.configure).not.toHaveBeenCalled();
  });

  it("should set additional configurations when the component mounts if options are passed through props", () => {
    props.options = {
      foo: "bar"
    };
    mount(<NextNProgress {...props} />);
    expect(NProgress.configure).toHaveBeenCalled();
  });

  it("should set the start position of the progress bar to the startPosition prop value when the routeChangeStart method is called", () => {
    wrapper.instance().routeChangeStart();
    expect(NProgress.set).toHaveBeenCalledWith(props.startPosition);
  });

  it("should start the progress bar when the routeChangeStart method is called", () => {
    wrapper.instance().routeChangeStart();
    expect(NProgress.start).toHaveBeenCalled();
  });

  it("should set the progress bar's done state when the routeChangeEnd method is called after the set delay", () => {
    jest.useFakeTimers();
    wrapper.instance().routeChangeEnd();
    expect(clearTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(
      expect.any(Function),
      props.stopDelayMs
    );
    expect(NProgress.done).not.toBeCalled();
    jest.runAllTimers();
    expect(NProgress.done).toHaveBeenCalledWith(true);
  });
});
