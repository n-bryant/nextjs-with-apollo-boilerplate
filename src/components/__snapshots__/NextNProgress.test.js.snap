// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextNProgress should render successfully 1`] = `
<JSXStyle
  dynamic={
    Array [
      "#fff",
      3,
      "#fff",
      "#fff",
    ]
  }
  id="1257616909"
>
  #nprogress{pointer-events:none;}#nprogress .bar{background:#fff;position:fixed;z-index:1031;top:0;left:0;width:100%;height:3px;}#nprogress .peg{display:block;position:absolute;right:0px;width:100px;height:100%;opacity:1;}#nprogress .spinner{display:"block";position:fixed;z-index:1031;top:19px;left:6px;}#nprogress .spinner-icon{width:18px;height:18px;box-sizing:border-box;border:solid 2px transparent;border-top-color:#fff;border-left-color:#fff;border-radius:50%;-webkit-animation:nprogresss-spinner 400ms linear infinite;-webkit-animation:nprogress-spinner 400ms linear infinite;animation:nprogress-spinner 400ms linear infinite;}.nprogress-custom-parent{overflow:hidden;position:relative;}.nprogress-custom-parent #nprogress .spinner,.nprogress-custom-parent #nprogress .bar{position:absolute;}@-webkit-keyframes nprogress-spinner{0%{-webkit-transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);}}@-webkit-keyframes nprogress-spinner{0%{-webkit-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}@keyframes nprogress-spinner{0%{-webkit-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}
</JSXStyle>
`;
